src/porepy/numerics/fv/biot.py

```diff
@@ -608,10 +611,344 @@ def compute_stress(self, g, u, data):
+        raise NotImplementedError(
+            """No discretize method implemented for the GradP
+                                  class. See the Biot class."""
+        )
+
+    def assemble_matrix_rhs(self, g, data):
+        """ Return the matrix and right-hand side for a discretization of the pressure
+        gradient term of the Biot equation.
+
+        Parameters:
+            g : grid, or a subclass, with geometry fields computed.
+            data: dictionary to store the data. For details on necessary keywords,
+                see method discretize()
+
+        Returns:
+            matrix: sparse csr (g.dim * g_num_cells, g.dim * g_num_cells) Discretization
+            matrix.
+            rhs: array (g.dim * g_num_cells) Right-hand side.
+        """
+        return self.assemble_matrix(g, data), self.assemble_rhs(g, data)
+
+    def assemble_matrix(self, g, data):
+        """ Return the matrix and right-hand side for a discretization of the pressure
+        gradient term of the Biot equation.
+
+        Parameters:
+            g (Grid): Computational grid, with geometry fields computed.
+            data (dictionary): With data stored.
+
+        Returns:
+            scipy.sparse.csr_matrix: System matrix of this discretization. The
+                size of the matrix will depend on the specific discretization.
+
+        Raises:
+            ValueError if the pressure gradient term has not already been discretized.
+        """
+        matrix_dictionary = data[pp.DISCRETIZATION_MATRICES][self.keyword]
+        if not "grad_p" in matrix_dictionary:
+            raise ValueError(
+                """GradP class requires a pre-computed discretization to be
+                             stored in the matrix dictionary."""
+            )
+        biot_alpha = data[pp.PARAMETERS][self.keyword]["biot_alpha"]
+        return matrix_dictionary["grad_p"] * biot_alpha
+
+    def assemble_rhs(self, g, data):
+        """ Return the zero right-hand side for a discretization of the pressure
+        gradient term.
+
+        @Runar: Is it correct that this is zero.
+
+        Parameters:
+            g (Grid): Computational grid.
+            data (dictionary): With data stored.
+
+        Returns:
+            np.ndarray: Zero right hand side vector with representation of boundary
+                conditions.
+        """
+        return np.zeros(self.ndof(g))
+
+
+class DivD(pp.numerics.mixed_dim.VectorEllipticDiscretization):
+    """ Class for the displacement divergence term of the Biot equation.
+    """
+
+    def ndof(self, g):
+        """ Return the number of degrees of freedom associated to the method.
+
+        In this case number of cells times dimension (stress dof).
+
+        Parameter
+        ---------
+        g: grid, or a subclass.
+
+        Return
+        ------
+        dof: the number of degrees of freedom.
+
+        """
+        return g.num_cells
+
+    def extract_displacement(self, g, solution_array, d):
```

Comment:
It seems a bit strange that all of these classes shall have an extract_displacement/pressure method.
It tells me that we still can make improvements in the code structure.

Reply:
I agree, this will hopefully evolve as we get more experience.